### ISSUE TITLE
Broken MIR: generator contains type ... after copy & destination propagation

### ISSUE THREAD

**Issue by tmiasko** [rust-lang/rust#76375]
x.rs:
```rust
#![crate_type = "lib"]

#[inline(always)]
pub fn f(s: bool) -> String {
    let a = "Hello world!".to_string();
    let b = a;
    let c = b;
    if s {
        c
    } else {
        String::new()
    }
}
```
y.rs
```rust
#![crate_type = "lib"]

pub async fn g() {
    x::f(true);
    h().await;
}

pub async fn h() {}
```

```console
$ rustc --edition=2018 -Zmir-opt-level=2 x.rs -Zunsound-mir-opts x.rs
$ rustc --edition=2018 -Zmir-opt-level=2 y.rs --extern x -L.
error: internal compiler error: src/librustc_mir/transform/generator.rs:752:13: Broken MIR: generator contains type std::string::String in MIR, but typeck only knows about {std::future::ResumeTy, impl std::future::Future, ()}
```

The copy propagation removes storage markers during optimization. Afterwards state transforms considers them to be always live across yield. Due to pass ordering, requires inliner to expose the issue.

**Comment by JohnTitor** [comment]
The ICE is no longer reproduced with the latest nightly on glacier, marking as `E-needs-test`.

**Comment by tmiasko** [comment]
For copy propagation this is still an issue, but the test case now requires `-Zunsound-mir-opts`.
For destination propagation the test case was fixed after changes that stop propagation if either local was ever borrowed. Hard to say if this fixed the issue more generally.

**Comment by JohnTitor** [comment]
@tmiasko Thanks for the heads-up!

Submitted a PR to readd this: rust-lang/glacier#479

**Comment by JohnTitor** [comment]
Triage: It's now compiled fine (again) with the latest nightly (I'd say it's finally fixed by #77306).
